dvc/remote/local/__init__.py

```diff
@@ -123,41 +125,65 @@ def makedirs(self, path_info):
         if not self.exists(path_info):
             os.makedirs(path_info.path)
 
-    @slow_link_guard
-    def link(self, cache_info, path_info):
-        cache = cache_info.path
-        path = path_info.path
+    def link(self, from_info, to_info, link_type=None):
+        from_path = from_info.path
+        to_path = to_info.path
 
-        assert os.path.isfile(cache)
+        assert os.path.isfile(from_path)
 
-        dname = os.path.dirname(path)
+        dname = os.path.dirname(to_path)
         if not os.path.exists(dname):
             os.makedirs(dname)
 
         # NOTE: just create an empty file for an empty cache
-        if os.path.getsize(cache) == 0:
-            open(path, "w+").close()
+        if os.path.getsize(from_path) == 0:
+            open(to_path, "w+").close()
 
-            msg = "Created empty file: {} -> {}".format(cache, path)
+            msg = "Created empty file: {} -> {}".format(from_path, to_path)
             logger.debug(msg)
             return
 
-        i = len(self.cache_types)
-        while i > 0:
-            try:
-                self.CACHE_TYPE_MAP[self.cache_types[0]](cache, path)
+        if not link_type:
+            self._default_link(from_info, to_info)
+        else:
+            link_method = self._get_link_method(link_type)
+            self._link(from_info, to_info, link_method)
+
+    @classmethod
+    def _get_link_method(cls, link_type):
+        try:
+            return cls.CACHE_TYPE_MAP[link_type]
+        except KeyError:
+            raise DvcException(
+                "Cache type: '{}' not supported!".format(link_type)
+            )
 
-                if self.protected:
-                    self.protect(path_info)
+    def _link(self, from_info, to_info, link_method):
+        if os.path.lexists(to_info.path):
+            raise DvcException(
+                "Link '{}' already exists!".format(to_info.path)
+            )
+        else:
+            link_method(from_info.path, to_info.path)
+        if self.protected:
+            self.protect(to_info)
+
+        msg = "Created {}'{}': {} -> {}".format(
+            "protected " if self.protected else "",
+            self.cache_types[0],
+            from_info.path,
+            to_info.path,
+        )
 
-                msg = "Created {}'{}': {} -> {}".format(
```

Comment:
But now we don't have this message for default link. Feels like _link is unnecessary duplication. Maybe you could simply pass a list of links to try out and not copy this code over?

Reply:
I do not agree that we don't have this message for default link. _link is called inside default link. Thought passing list of links to try out seems like better idea than creating default_link method, which sounds at least forced.

Reply:
Ah, missed that. Sorry.